onnx/defs/math/defs.cc

```diff
@@ -3660,4 +3660,534 @@ ONNX_OPERATOR_SET_SCHEMA(
+                "Input tensor representing a real or complex valued signal. "
+                "For real input, the following shape is expected: [batch_size][n_fft]. "
+                "For complex input, the following shape is expected: [batch_size][n_fft][2], where " 
+                "[batch_size][n_fft][0] represents the real component and [batch_size][n_fft][1] represents the imaginary component of the signal.",
+                "T",
+                OpSchema::Single,
+                true,
+                1,
+                OpSchema::NonDifferentiable)
+            .Output(
+                0,
+                "output",
+                "The inverse 1-dimensional discrete Fourier transform of the input. "
+                "The output tensor has the same shape and type as the input.",
+                "T",
+                OpSchema::Single,
+                true,
+                1,
+                OpSchema::NonDifferentiable)
+            .TypeConstraint(
+                "T",
+                {"tensor(float16)", "tensor(float)", "tensor(double)"},
+                "Constrain input and output types to float tensors.")
+            .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+                propagateElemTypeFromInputToOutput(ctx, 0, 0);
+                // To support multidimensional DFT the signal_ndim may != 1
+                const int64_t batch_ndim = 1;
+                const int64_t component_ndim = 1;
+                
+                // To support multidimensional DFT the signal_ndim may != 1,
+                // but this attribute is not included on the schema yet
+                const auto signal_ndim = 1;
+
+                // Add 1 dimensions to the rank for the batch size dimension
+                auto real_inputs_ndim = signal_ndim + batch_ndim;
+                // Add 2 dimensions to the rank for the batch size dimension AND component dimension
+                auto complex_inputs_ndim = signal_ndim + batch_ndim + component_ndim;
+
+                auto& input_shape = getInputShape(ctx, 0);
+                ONNX_NAMESPACE::TensorShapeProto result_shape = input_shape;
+
+                auto dim_size = static_cast<int64_t>(input_shape.dim_size());
+                if (dim_size == real_inputs_ndim) {
+                    // For real input: the output is same shape, but with extra dim to represent complex output (real/imaginary)
+                    result_shape.add_dim()->set_dim_value(2);
+                } else if (dim_size == complex_inputs_ndim) {
+                    // For complex input: do nothing
+                } else {
+                  fail_shape_inference(
+                        "The input_shape must [batch_size][n_fft] for real values or [batch_size][n_fft][2] for complex values.")
+                }
+
+                updateOutputShape(ctx, 0, result_shape);
+            }));
+
+static const char* HannWindow_ver13_doc =
+    R"DOC(Generate a Hann window.)DOC";
```

Comment:
Can we add any link to help explain what a Hann window is?

Reply:
If all the window ops are mostly identical, I would suggest factoring it out to avoid duplication. This will help making updates/changes while keeping them in sync. (E.g., see https://github.com/onnx/onnx/blob/5282dd787b2c913110c26f34de7b624cf7ab9c01/onnx/defs/math/defs.cc#L53 )